src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb

```diff
@@ -847,5 +847,139 @@ BC37329: A value of type 'System.Threading.Lock' is not supported in SyncLock. C
+
+    Public Class DerivedLock
+        Inherits Lock
+        Implements IDerivedLock
+    End Class
+
+    Interface IDerivedLock
+    End Interface
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics(
+"BC37329: A value of type 'System.Threading.Lock' is not supported in SyncLock. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.
+        SyncLock l2 ' 1
+                 ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        Dim l3 As DerivedLock = CType(l2, DerivedLock) ' 2
+                                      ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        l3 = DirectCast(l2, DerivedLock) ' 3
+                        ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        l3 = TryCast(l2, DerivedLock) ' 4
+                     ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        Dim l4 As IDerivedLock = CType(l2, IDerivedLock) ' 5
+                                       ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        l4 = DirectCast(l2, IDerivedLock) ' 6
+                        ~~
+BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+        l4 = TryCast(l2, IDerivedLock) ' 7
+                     ~~
+")
+        End Sub
+
+        <Fact>
+        Public Sub LockType_Derived_Execution()
+            Dim source = <![CDATA[
+Imports System
+Imports System.Threading
+
+Module Program
+    Sub Main()
+        Dim l1 As DerivedLock = New DerivedLock()
+        Dim l2 As Lock = l1
+        Dim l3 As DerivedLock = CType(l2, DerivedLock)
+        SyncLock l3
+            Console.WriteLine("locked")
+        End SyncLock
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Lock
+    End Class
+
+    Public Class DerivedLock
+        Inherits Lock
+        Implements IDerivedLock
+    End Class
+
+    Interface IDerivedLock
+    End Interface
+End Namespace
+]]>.Value
+            Dim comp = CreateCompilation(source, options:=TestOptions.ReleaseExe)
+            Dim verifier = CompileAndVerify(comp, expectedOutput:="locked")
+            ' (9, 39) : warning BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+            verifier.VerifyDiagnostics(Diagnostic(ERRID.WRN_ConvertingLock, "l2").WithLocation(9, 39))
```

Comment:
>VerifyDiagnostics(Diagnostic(ERRID.WRN_ConvertingLock, "l2").WithLocation(9, 39)) [](http://example.com/codeflow?start=21&length=81)

Please use `AssertTheseDiagnostics` style helpers to verify diagnostics for VB

Reply:
It's not available for the verifier it seems.

Reply:
> It's not available for the verifier it seems.

What prevents us from making it available?

At the very list we should be able to get diagnostics from the verifier and call this helper.

```
    ''' <param name="errors"></param>
    ''' <param name="errs">Expected errors according to this schema
    ''' &lt;expected&gt;[full errors text]&lt;/expected&gt;</param>
    ''' <param name="suppressInfos">True to ignore info-severity diagnostics.</param>
    ''' <remarks></remarks>
    <Extension()>
    Public Sub AssertTheseDiagnostics(errors As ImmutableArray(Of Diagnostic), errs As XElement, Optional suppressInfos As Boolean = True)
        If errs Is Nothing Then
            errs = <errors/>
        End If
        Dim expectedText = FilterString(errs.Value)
        AssertTheseDiagnostics(errors, expectedText, suppressInfos)
    End Sub
```